Core Working Group call notes: October 11, 2024 (#4198)
tock/tock: doc/wg/core/notes/core-notes-2024-10-11.md

 * Amit: In userspace, the subtle thing is that when we iterate through outstanding alarms, we use the current counter value to do comparisons, rather than when the alarm was scheduled to fire. That is correct because now might proceed as we iterate, but on the other hand if the timer overflows and I had a long timer, all of a sudden this timer appears to not have fired. For example, if the reference is 2, firing is 1, and current time is 3. It looks like it's brand new instead of recently expired. I think we can check for this, but it needs more testing.
 * Tyler: What's your use case that's causing you to chase down these bugs?
 * Amit: I've been working on the SMA_Q3 smartwatch platform. Literally writing a clock right now with adjustments based on the GPS. I've still not really set really long wrapping timers, but I encountered a different bug where inserting ordering was wrong. Then I was writing more tests and found this too.
 * Leon: I thought my tests in the kernel covered every case. Did I miss one?
 * Amit: I suspect it was just missed. And I'll add one for it
 * Brad: Are those tests on LiteX?
 * Leon: No, they're capsule tests
 * Brad: And those have smaller sizes?

>  * Leon: 32, 24, and 16. Came out of the major rewrite where I ripped out the timer logic from the userspace boundary

Reviewer: ```suggestion
 * Leon: 32, 24, and 64. Came out of the major rewrite where I ripped out the timer logic from the userspace boundary
```

 * Brad: So, now I've been looking into machinery for components to standardize configuring boards. Still iterating on this with new ideas.
 * Leon: I have been trying to make it so the core kernel logic doesn't need to care about process-specific debugging info. That created so many issues requiring generic types literally everywhere in the codebase. It's still worth pursuing disentangling process debugging and the core kernel, but this PR is still a great first step.
 * Brad: What I've found is that types are tractable as long as everything is for ProcessStandard, but not for other things. That does make different implementations more complicated, but simplifies most of our code
### Nightly Update
 * https://github.com/tock/tock/pull/4193
 * Brad: This PR has some static mut issues. Amit says they're "easy to handle" and he'll help.
 * Hudson: We could also make a partial jump if we needed, to somewhere less than here

>  * Leon: The warnings in this nightly are good, but they haven't reached the level where there are a TON of them. And they occur in tests

Reviewer: ```suggestion
 * Leon: The warnings in this nightly are not too bad, but they haven't gotten the point of checking board-tests, where there's a ton of mutable statics still
```

 * Ben: The board harness has architecture, path in the Tock repo, ability to get UART port, and functions for flashing it
 * Ben: The testing framework such as the "c_hello.py" file, has a test, steps to load it on the board such as flashing the kernel and app or apps. Then for oneshot tests, which are the typical ones, they run once and collect info
 * Ben: Two types of tests, AnalyzeConsole expects output from serial ports. There's also a WaitForConsoleMessage test which instead just waits for a single message.
 * Ben: This can run on real hardware with Treadmill now. In the test-execute for Treadmill, the runner will load things, wait for a message, and then return success/failure
 * Leon: The basic idea here is that for any given test we want to combine a specification for interaction with a board with a specification for how the test works. So the board is interaction and the test is the strategy of when to flash things, when to flush buffers, when to expect output, when to use GPIO pins, etc. We can inherit parts of this to abstract away common choices here that many tests use. For example WaitForConsoleMessage is an abstraction that just takes an app and a string.

>  * Leon: So the goal here was to balance expressibility and abstracting detail away for simple tasks

Reviewer: ```suggestion
 * Leon: So the goal here was to balance expressiveness and abstracting detail away for simple tasks
```
Probably misspoke there. :smile:

Author: In my head the two words mean exactly the same thing. So I'd blame me for this one.
